### ISSUE TITLE
When to stop training , how to avoid overfitting and How to evaluate weights 

### ISSUE THREAD

**Issue by mwindowshz** [AlexeyAB/darknet#291]
Hi
I am running training on VOC data set, (all images from 2007 and 2012 as explained on md file)
using the yolo_voc.2.0.cfg (without random)

the learning avg loss seems to get stuck at 0.17-0.2
and does not go lower, 
I ran recall on several weights, and could not figure out, what is happening,
should I have stopped training earlier, should I continue? should I change a parameter?

what is actually happening?

```
100: 2.151674, 2.241611 avg, 0.001000 rate, 2.095000 seconds, 6400 images
200: 1.562922, 1.542185 avg, 0.001000 rate, 2.095000 seconds, 12800 images
300: 1.112201, 1.370037 avg, 0.001000 rate, 2.064000 seconds, 19200 images
400: 1.014449, 1.306595 avg, 0.001000 rate, 2.088000 seconds, 25600 images
500: 1.140327, 1.173284 avg, 0.001000 rate, 2.065000 seconds, 32000 images
600: 1.325276, 1.234068 avg, 0.001000 rate, 2.045000 seconds, 38400 images
700: 0.937698, 1.076297 avg, 0.001000 rate, 1.974000 seconds, 44800 images
800: 1.091687, 1.148948 avg, 0.001000 rate, 1.958000 seconds, 51200 images
900: 0.971230, 0.958444 avg, 0.001000 rate, 2.162000 seconds, 57600 images
1000: 0.670229, 0.988479 avg, 0.001000 rate, 2.073000 seconds, 64000 images
2000: 0.724700, 0.759144 avg, 0.001000 rate, 2.068000 seconds, 128000 images
3000: 0.598594, 0.625907 avg, 0.001000 rate, 2.081000 seconds, 192000 images
4000: 0.449722, 0.572213 avg, 0.001000 rate, 2.065000 seconds, 256000 images
5000: 0.595720, 0.541806 avg, 0.001000 rate, 2.038000 seconds, 320000 images
6000: 0.539147, 0.486868 avg, 0.001000 rate, 2.094000 seconds, 384000 images
7000: 0.460493, 0.453343 avg, 0.001000 rate, 2.066000 seconds, 448000 images
8000: 0.644437, 0.450894 avg, 0.001000 rate, 2.082000 seconds, 512000 images
9000: 0.338761, 0.455005 avg, 0.001000 rate, 2.051000 seconds, 576000 images
10000: 0.253477, 0.386221 avg, 0.001000 rate, 2.063000 seconds, 640000 images
11000: 0.404973, 0.386226 avg, 0.001000 rate, 2.103000 seconds, 704000 images
12000: 0.329849, 0.350028 avg, 0.001000 rate, 2.057000 seconds, 768000 images
13000: 0.254458, 0.347430 avg, 0.001000 rate, 2.039000 seconds, 832000 images
14000: 0.484491, 0.347095 avg, 0.001000 rate, 2.072000 seconds, 896000 images
15000: 0.365126, 0.308175 avg, 0.001000 rate, 2.035000 seconds, 960000 images
16000: 0.219049, 0.306167 avg, 0.001000 rate, 2.073000 seconds, 1024000 images
17000: 0.468699, 0.336137 avg, 0.001000 rate, 2.041000 seconds, 1088000 images
18000: 0.423874, 0.315608 avg, 0.001000 rate, 2.051000 seconds, 1152000 images
19000: 0.296636, 0.284299 avg, 0.001000 rate, 2.079000 seconds, 1216000 images
20000: 0.274259, 0.311331 avg, 0.001000 rate, 2.051000 seconds, 1280000 images
21000: 0.302300, 0.272038 avg, 0.001000 rate, 2.069000 seconds, 1344000 images
22000: 0.234984, 0.256759 avg, 0.001000 rate, 2.054000 seconds, 1408000 images
23000: 0.206957, 0.239507 avg, 0.001000 rate, 2.039000 seconds, 1472000 images
24000: 0.310842, 0.252890 avg, 0.001000 rate, 2.053000 seconds, 1536000 images
25000: 0.226618, 0.254046 avg, 0.000100 rate, 2.059000 seconds, 1600000 images
26000: 0.203973, 0.231458 avg, 0.000100 rate, 2.069000 seconds, 1664000 images
27000: 0.278905, 0.232688 avg, 0.000100 rate, 2.021000 seconds, 1728000 images
28000: 0.282352, 0.195806 avg, 0.000100 rate, 2.068000 seconds, 1792000 images
29000: 0.141398, 0.196842 avg, 0.000100 rate, 2.066000 seconds, 1856000 images
30000: 0.189394, 0.196730 avg, 0.000100 rate, 2.053000 seconds, 1920000 images
31000: 0.203255, 0.197613 avg, 0.000100 rate, 2.054000 seconds, 1984000 images
32000: 0.164793, 0.191329 avg, 0.000100 rate, 2.058000 seconds, 2048000 images
33000: 0.242713, 0.196200 avg, 0.000100 rate, 2.094000 seconds, 2112000 images
34000: 0.234014, 0.183112 avg, 0.000100 rate, 2.072000 seconds, 2176000 images
35000: 0.181141, 0.179613 avg, 0.000010 rate, 2.050000 seconds, 2240000 images
36000: 0.177562, 0.207329 avg, 0.000010 rate, 2.037000 seconds, 2304000 images
37000: 0.181070, 0.180319 avg, 0.000010 rate, 2.068000 seconds, 2368000 images
38000: 0.124009, 0.169776 avg, 0.000010 rate, 2.082000 seconds, 2432000 images
39000: 0.161391, 0.183015 avg, 0.000010 rate, 2.031000 seconds, 2496000 images
40000: 0.113188, 0.179209 avg, 0.000010 rate, 2.032000 seconds, 2560000 images
41000: 0.147263, 0.193683 avg, 0.000010 rate, 2.029000 seconds, 2624000 images
42000: 0.142396, 0.181724 avg, 0.000010 rate, 2.014000 seconds, 2688000 images
43000: 0.103506, 0.159163 avg, 0.000010 rate, 2.032000 seconds, 2752000 images
44000: 0.212113, 0.171423 avg, 0.000010 rate, 2.034000 seconds, 2816000 images
45000: 0.293601, 0.186927 avg, 0.000010 rate, 2.033000 seconds, 2880000 images
46000: 0.149545, 0.189659 avg, 0.000010 rate, 2.053000 seconds, 2944000 images
47000: 0.185092, 0.180458 avg, 0.000010 rate, 2.021000 seconds, 3008000 images
48000: 0.144500, 0.170354 avg, 0.000010 rate, 2.074000 seconds, 3072000 images
49000: 0.113062, 0.196493 avg, 0.000010 rate, 2.065000 seconds, 3136000 images
50000: 0.156577, 0.174924 avg, 0.000010 rate, 2.117000 seconds, 3200000 images
51000: 0.136423, 0.159352 avg, 0.000010 rate, 2.062000 seconds, 3264000 images
52000: 0.248675, 0.200785 avg, 0.000010 rate, 2.074000 seconds, 3328000 images
53000: 0.240793, 0.201979 avg, 0.000010 rate, 2.055000 seconds, 3392000 images
54000: 0.209582, 0.179752 avg, 0.000010 rate, 2.063000 seconds, 3456000 images
55000: 0.169098, 0.176866 avg, 0.000010 rate, 2.067000 seconds, 3520000 images
56000: 0.114330, 0.197297 avg, 0.000010 rate, 2.067000 seconds, 3584000 images
57000: 0.175227, 0.189928 avg, 0.000010 rate, 2.064000 seconds, 3648000 images
58000: 0.128261, 0.159543 avg, 0.000010 rate, 2.051000 seconds, 3712000 images
59000: 0.221510, 0.176424 avg, 0.000010 rate, 2.050000 seconds, 3776000 images
60000: 0.140829, 0.171079 avg, 0.000010 rate, 2.072000 seconds, 3840000 images
```

you can see that from 30000 or 35000 it does not get lower.
**Testing Recall**
10000
recall
  4951  8962 12032       RPs/Img: 3.12   IOU: 61.63%     Recall:74.48%
20000
 4951  9118 12032       RPs/Img: 3.00   IOU: 63.22%     Recall:75.78%
30000
 4951  9270 12032       RPs/Img: 2.82   IOU: 64.47%     **Recall:77.04%**
40000
 4951  9253 12032       RPs/Img: 2.83   IOU: 64.61%     Recall:76.90%
50000
 4951  9257 12032       RPs/Img: 2.82   IOU: 64.62%     Recall:76.94%
60000
 4951  9269 12032       RPs/Img: 2.81   IOU: 64.60%     **Recall:77.04%**

Recall at 30000 and 60000 is the same! and IOU is just slightly different.

So basically, should I have stopped training earlier?
do I have to set the cfg file diffrent? maybe after 10000 iteration need to change some parameter such as 
momentum or decay?

![image](https://user-images.githubusercontent.com/17617515/33539637-1d6f6c92-d8d0-11e7-88d0-6c94bf5e3d17.png)

Would appreciate any explanation for this situation.
Thanks in advance
M.

**Comment by springkim** [comment]
You can measure loss by get_cost function(i am not sure) .

**Comment by sivagnanamn** [comment]
@springkim Thank you for the pointer. Could you please elaborate why validation loss shouldn't be measured during training?
